include/unifex/stop_when.hpp

```diff
@@ -34,359 +34,372 @@
-      std::optional<typename stop_token_type_t<
-          Receiver>::template callback_type<cancel_callback>>
-          stopCallback_;
-      UNIFEX_NO_UNIQUE_ADDRESS result_variant result_;
-      UNIFEX_NO_UNIQUE_ADDRESS connect_result_t<Source, source_receiver> sourceOp_;
-      UNIFEX_NO_UNIQUE_ADDRESS
-      connect_result_t<Trigger, trigger_receiver> triggerOp_;
-    };
-
-    template <typename Source, typename Trigger>
-    struct _sndr {
-      class type;
-    };
-
-    template <typename Source, typename Trigger>
-    using stop_when_sender = typename _sndr<Source, Trigger>::type;
-
-    template <typename Source, typename Trigger>
-    class _sndr<Source, Trigger>::type {
-      using stop_when_sender = type;
-
-      template <typename... Values>
-      using decayed_type_list = type_list<type_list<std::decay_t<Values>...>>;
-
-      template <
-          template <typename...> class Outer,
-          template <typename...> class Inner>
-      struct compose_nested {
-        template <typename... Lists>
-        using apply = Outer<typename Lists::template apply<Inner>...>;
-      };
-
-    public:
-      template <
-          template <typename...> class Variant,
-          template <typename...> class Tuple>
-      using value_types = typename sender_traits<Source>::
-          template value_types<concat_type_lists_unique_t, decayed_type_list>::
-              template apply<compose_nested<Variant, Tuple>::template apply>;
-
-      template <template <typename...> class Variant>
-      using error_types =
-          typename concat_type_lists_unique_t<
-              sender_error_types_t<Source, decayed_tuple<type_list>::template apply>,
-              type_list<std::exception_ptr>>::template apply<Variant>;
-
-      static constexpr bool sends_done = true;
-
-      template <typename Source2, typename Trigger2>
-      explicit type(Source2&& source, Trigger2&& trigger) noexcept(
-          std::is_nothrow_constructible_v<Source, Source2> &&
+    }
+
+  private:
+    type* op_;
+  };
+
+  void notify_source_complete() noexcept { this->notify_trigger_complete(); }
+
+  void notify_trigger_complete() noexcept {
+    stopSource_.request_stop();
+    if (activeOpCount_.fetch_sub(1, std::memory_order_acq_rel) == 1) {
+      stopCallback_.reset();
+      deliver_result();
+    }
+  }
+
+  void deliver_result() noexcept {
+    UNIFEX_TRY {
+      std::visit(
+          [this](auto&& tuple) {
+            if constexpr (
+                std::tuple_size<
+                    std::remove_reference_t<decltype(tuple)>>::value != 0) {
+              std::apply(
+                  [&](auto set_xxx, auto&&... args) {
+                    set_xxx(
+                        std::move(receiver_),
+                        static_cast<decltype(args)>(args)...);
+                  },
+                  static_cast<decltype(tuple)>(tuple));
+            } else {
+              // Should be unreachable
+              std::terminate();
```

Comment:
This kinda bugs me every time I see it; I feel like we should change the `if constexpr` to a `static_assert` and delete this branch.

Reply:
#505

Reply:
This branch is there because one of the variant's alternative types is the empty `tuple<>` and `std::visit` is going to instantiate the lambda with that type as an argument.
This type doesn't have a `set_xxx` cpo type and so can't be passed to the `std::apply` call.

So there is a valid, reachable code-path according to the compiler, it's just that you should never be calling `deliver_result()` at runtime when the variant holds the type `std::tuple<>`.